shared/docs/flows/AppHash.md

```diff
@@ -1,40 +1,92 @@
-# AppHash
+# AppHash <!-- omit in toc -->
+
+- [Context Initialization](#context-initialization)
+- [Block Application](#block-application)
+- [Block Commit](#block-commit)
 
 ## Context Initialization
 
+This flow shows the process of context initialization between all the modules requirer to apply a block and compute a state hash during the consensus lifecycle.
+
 ```mermaid
 sequenceDiagram
     %% autonumber
     participant N as Node
     participant C as Consensus
     participant U as Utility
     participant P as Persistence
-    participant PP as Persistence (PostgresDB)
-    participant PM as Persistence (MerkleTree)
     participant P2P as P2P
 
-    %% Should this be P2P?
     N-->>C: HandleMessage(anypb.Any)
     critical NewRound Message
         C->>+U: NewContext(height)
         U->>P: NewRWContext(height)
         P->>U: PersistenceRWContext
-        U->>U: Store persistenceContext
+        U->>U: store context<br>locally
         U->>-C: UtilityContext
-        C->>C: Store utilityContext
-        Note over C, PM: See 'Block Application'
+        C->>C: store context<br>locally
+        Note over C, P: See 'Block Application'
     end
+
     Note over N, P2P: Hotstuff lifecycle
     N-->>C: HandleMessage(anypb.Any)
+
     critical Decide Message
-        Note over C, PM: See 'Block Commit'
+        Note over C, P: See 'Block Commit'
     end
 ```
 
 ## Block Application
 
-TODO(olshansky): Add a sequenceDiagram here.
+```mermaid
+sequenceDiagram
+    participant C as Consensus
+    participant U as Utility
+    participant P as Persistence
+
+    alt as leader
+        C->>+U: GetProposalTransactions(proposer, maxTxBz, [lastVal])
+        U->>U: reap mempool
+        U->>-C: txs
+        Note over C, U: fallthrough to replica behaviour
+    else as replica
+        C->>+U: ApplyBlock(height, proposer, txs, lastVals)
+        loop for each operation in tx
+            U->>P: Get*/Set*
+            P->>U: result, err_code
+            U->>U: validation<br>logic
+            U->>P: StoreTransaction(tx)
+            P->>P: store tx<br>locally
+            P->>U: result, err_code
+        end
+        U->>+P: UpdateAppHash()
+        Note over P: Update State Hash
+        P->>-U: stateHash
+        U->>-C: stateHash
+    end
```

Comment:
There are a few observations to the diagram, and a suggestion to accompany them.

Given [the conversation above](https://github.com/pokt-network/pocket/pull/252/files#r990511395):
* it seems the activity around getting the proposal transactions is optional (when acting as leader) and does not exclude the rest of the sequence. I put the conditional section in an `opt` block instead of an `alt` block.
* I reflected my understanding of @deblasis' comment with regards to the `reap mempool` message by making it mutually exclusive with `get QC`. I don't know where the QC comes from, but reflected my guess that utility might have it.

Other changes: 
* The loop label might have been meant to be `tx in txs` instead of `operations in tx`.
* When actors message themselves into doing work, they should be activated
* Responses should be dotted lines.

```suggestion
    opt if leader
        C->>U: GetProposalTransactions(proposer, maxTxBz, [lastVal])
        activate U
        alt no quorum for Prepare
        U->>U: reap mempool
        activate U
        deactivate U
        else
        U->>U: get QC
        activate U
        deactivate U
        end
        U-->>C: txs
        deactivate U
    end
    C->>+U: ApplyBlock(height, proposer, txs, lastVals)
    loop for each tx in txs
        U->>+P: Get*/Set*
        P-->>-U: result, err_code
        U->>U: Validation logic
        activate U
        deactivate U
        U->>+P: StoreTransaction(tx)
        P->>P: Store tx locally
        activate P
        deactivate P
        P-->>-U: result, err_code
    end
    U->>+P: UpdateAppHash()
    P->>P: Update state hash
    activate P
    deactivate P
    P-->>-U: stateHash
    U-->>-C: stateHash
```

Reply:
[Here's a link](https://mermaid-js.github.io/mermaid-live-editor/edit#pako:eNqdVMtu2zAQ_JUFT04jI3cdDDQq0BxVOOqlKoqFtLaIUiRDrgK7Qf69S8mvRkZQRAdC2MfM7HClF9W4llSuIj0NZBv6onEbsK8tyOMxsG60R8tQAEYonI1k4xDn-SrlK9ZG836eLVO2pBB15ERT26nGeQa9AUPYUphC6SmWq1WVw1fiMjjvIprHgDZiw1oULPwYpJBBj7vH3f2fDH4YjPwdzc-bM0oqf0YmqC5ihsE6eBpcGHrYuABlIFFK55JqIg-EHnrqvXPmfcyWrkXJxDnolhi-FR-Cs-0lmsAVOfAuvtd36kl-3gr9Z-_N_t645veiI73tOIOzlwKWwcHGeLDROOdHlwS7kwrQ9l_SNNdtOV7Vp7u1HOdUmUQuRyfjYISKQviV1m3mihDqFtPdCuFWNx_x5yhkzS7QxbYseHdzqWm1OhalccQKNGZ_hbC8Slj-93gn74_CKi8jktzAA8ZucdB00DPlIHI6O8lP2be0cyEnEWPrw6lzXJBlMYurTPUUetStfPMvKVYr7qinWuXy2tIGZZZa1fZVSnFgt97bRuUcBsrUMMo8_CJUvpE9ode_LdtK7Q) to a rendered version of this suggestion

Reply:
Nice one, the "replica" behaviour (and the fallback to replica from the leader) looks clearer as well IMHO

Reply:
Good feedback @iajrz gives me more confidence moving forward with sequence diagrams in general. Also valuable edits IMO

Reply:
Nicely done! @andrewnguyen22 do you still want @oten91 eyes on this before we close it out?

Reply:
I wouldn't say it's required for my approval. I am awaiting Olshansk's changes / response to Irving's however

Reply:
It's important to note - `StoreTransaction()` isn't called until after the `Commit()` phase along side `StoreBlock()`. Please adjust accordingly

Reply:
@iajrz The suggestion is 💯 !